### PR TITLE
506 make custom split fun

### PR DESCRIPTION
**Pull request by gmbecker**
This PR has the newly implement make_split_fun and related functions in it, including tests and full function documentation.

### CODE REVIEW

**Comment by wwojciech** [comment]
Hi @gmbecker 

I am getting this below warning when trying to load the package from the source code (my R version is 4.2.2):
```
> devtools::load_all(".")
ℹ Loading rtables
Loading required package: magrittr
Loading required package: formatters
Creating a generic function for ‘nrow’ from package ‘base’ in package ‘rtables’
Creating a generic function for ‘ncol’ from package ‘base’ in package ‘rtables’
Creating a generic function for ‘rbind’ from package ‘base’ in package ‘rtables’
Warning message:
Objects listed as exports, but not present in namespace:
• [
• [<- 
```

File: R/make_split_fun.R [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+#' @param spl Split.
+#' @param ... dots.
+#' @export
+#' @seealso make_split_fun
+#' @family make_custom_split
+drop_facet_levels <- function(df, spl, ...) {
+    if(!is(spl, "VarLevelSplit") || is.na(spl_payload(spl)))
+        stop("Unable to determine faceting variable in drop_facet_levels application.")
+    var <- spl_payload(spl)
+    df[[var]] <- factor(df[[var]])
```
Comment (wwojciech) [inline R/make_split_fun.R:292]: out of curiosity, why not
```suggestion
    df[[var]] <- droplevels(df[[var]])
```

File: R/make_split_fun.R [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+                  function(nm) c(splres[[nm]], newstuff[[nm]]))
+    names(ret) <- names(splres)
+    .fixupvals(ret)
+}
+
+.can_take_spl_context <- function(f) any(c(".spl_context", "...") %in% names(formals(f)))
+
+#' Create a Custom Splitting Function
+#'
+#' @param pre list. Zero or more functions which operate on the incoming data and return a new data frame that should split via `core_split`. They will be called on the data in the order they appear in the list
```
Comment (wwojciech) [inline R/make_split_fun.R:82]: so the output of the previous function will be the `df` argument of the next function on the list?

**Comment by Melkiades** [comment]
> I am getting this below warning when trying to load the package from the source code (my R version is 4.2.2):
```
> devtools::load_all(".")
ℹ Loading rtables
Loading required package: magrittr
Loading required package: formatters
Creating a generic function for ‘nrow’ from package ‘base’ in package ‘rtables’
Creating a generic function for ‘ncol’ from package ‘base’ in package ‘rtables’
Creating a generic function for ‘rbind’ from package ‘base’ in package ‘rtables’
Warning message:
Objects listed as exports, but not present in namespace:
• [
• [<- 
```
This happens always when using devtools

**Comment by gmbecker** [comment]
> > I am getting this below warning when trying to load the package from the source code (my R version is 4.2.2):
> 
> ```
> > devtools::load_all(".")
> ℹ Loading rtables
> Loading required package: magrittr
> Loading required package: formatters
> Creating a generic function for ‘nrow’ from package ‘base’ in package ‘rtables’
> Creating a generic function for ‘ncol’ from package ‘base’ in package ‘rtables’
> Creating a generic function for ‘rbind’ from package ‘base’ in package ‘rtables’
> Warning message:
> Objects listed as exports, but not present in namespace:
> • [
> • [<- 
> ```
> 
> This happens always when using devtools

This is a false positive from devtools, you can ignore it

File: R/make_split_fun.R (hunk 2 above)
Comment (gmbecker) [reply R/make_split_fun.R:82]: @wwojciech correct, each `pre` function modifies `df` and provides the input for the next one in the list.

**Comment by clarkliming** [comment]
no impact for downstream 👍

**Comment by shajoezhu** [comment]
blocking this until we checked in scda.test repo

**Review by Melkiades — APPROVED** [review]
It is very clear and well-documented in my opinion. Also, it is a nice addition with a well-contained doc page with separated function documentation and a good family structure (`#' @family make_custom_split`).

I think it is important though to add a list of currently supported functions that can work with `make_split_fun` and not (via lists in the details section with the 3 options, pre, core, post) and seealso in other fun-doc pages. It is atm a bit self-standing. For example, it would be interesting to have better clarity on where I should use `trim_levels_in_group` and `trim_levels_in_facets`. The latter is missing an example; I know it is straightforward but I think it is needed with some disclaimers specifying why it is different from other trims.

File: man/make_split_fun.Rd [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+of the same structure specified above for core splitting.
+
+In both the pre- and  post-processing cases, multiple functions can
+be specified. When this happens,  they are applied sequentially, in
+the order they  appear in the list passed to  the relevant argument
+(\code{pre} and \code{post}, respectively).
+}
+\examples{
+mysplitfun <- make_split_fun(pre = list(drop_facet_levels),
+                             post = list(add_overall_facet("ALL", "All Arms")))
```
Comment (shajoezhu) [inline man/make_split_fun.Rd:76]: can me make a second example please?

mysplitfun2, where the list include two functions?

**Review by shajoezhu — APPROVED** [review]
lgtm! Thanks @gmbecker minor suggestion for adding more examples with two functions in the list

**Comment by shajoezhu** [comment]
Downstream is all checked. This PR is good to go. Let's address the new example in a sperate issue.

**Pull request merged**